smart_simulation/scale.py

```diff
@@ -0,0 +1,235 @@
+        freq=frequency,
+    )
+
+    upsampled_data = pd.DataFrame(index=upsampled_time_stamps).assign(servings=0)
+
+    for day in dataset.index:
+        for serving in range(dataset.loc[day]["servings"]):
+            period = random.choices(population=time_stamps, weights=weights, k=1)[0]
+            time_stamp = day + pd.to_timedelta(period)
+            upsampled_data.at[time_stamp, "servings"] += 1
+
+    return upsampled_data
+
+
+def create_weight_data(dataset: pd.DataFrame, arrival_template: str) -> pd.DataFrame:
+    """
+    Create weight data based on consumption and shipment arrivals
+    Args:
+        dataset: Dataset of time stamps and servings (consumed) by consumer
+        arrival_template: Arrival windows and probabilities used for random selection of new product arrival on scale
+
+    Returns:
+        Pandas DataFrame of weight measurements
+
+    """
+    arrival_options = auxiliary.arrival_options[arrival_template]
+    arrival_options_weights = tuple(
+        auxiliary.arrival_options_weights[arrival_template].values()
+    )
+    stock_weight = auxiliary.stock_weight
+    quantity_to_weight = auxiliary.quantity_to_weight
+    data = dataset.copy()
+    data = data.assign(weight=0.0)
+    weight_col = data.columns.get_loc("weight")
+    servings_col = data.columns.get_loc("servings")
+
+    data.iat[0, weight_col] = stock_weight
+
+    time_step = 1
+    while time_step < len(data.index):
+        previous_step_weight = data.iat[time_step - 1, weight_col]
+        consumption_servings = data.iat[time_step, servings_col]
+
+        if previous_step_weight == 0.0:
+            resupply_arrival = random.choices(
+                population=[*arrival_options], weights=arrival_options_weights, k=1
+            )[0]
+            arrival_timing = arrival_options[resupply_arrival][0](
+                *arrival_options[resupply_arrival][1]
+            )
+            time_step = time_step + arrival_timing
+            if time_step > len(data.index) + 1:
+                continue
+            if time_step < 1:
+                time_step = time_step - arrival_timing
+                continue
+
+            data.iat[time_step, weight_col] += stock_weight
+            time_step += 1
+            previous_step_weight = data.iat[time_step - 1, weight_col]
+            consumption_servings = data.iat[time_step, servings_col]
+
+        if consumption_servings == 0:
+            data.iat[time_step, weight_col] = data.iat[time_step - 1, weight_col]
+
+        serving_weight = quantity_to_weight[0](*quantity_to_weight[1])
+        consumption_weight = serving_weight * consumption_servings
+
+        if consumption_weight > previous_step_weight:
+            data.iat[time_step, weight_col] = 0.0
+        else:
+            data.iat[time_step, weight_col] = previous_step_weight - consumption_weight
+
+        time_step += 1
+
+    return data
+
+
+def calibration_error(
```

Comment:
I would note in the docstring that this function edits the dataframe in place (contrast with `create_weight_data()` above that edits and returns a copy of the data). Best to stick with one or the other paradigm - inplace or copy.

Reply:
Intention for copy in `create_weight_data` is to be able to create weight data based off various arrival scenarios / templates. Leave the upsampled servings dataset alone so it can be reused. I will pick one or the other and change.

When creating multiple datasets I plan to save multiple CSVs: `upsampled_servings`, `weight_measurements`, `modified_weight_measurements`. This will allow me to reference back to a **ground truth**.